### ISSUE TITLE
Add custom loss

### ISSUE THREAD

**Issue by jthsieh** [uber/pyro#1059]
I've been playing around with Pyro and it has worked pretty well so far. Thanks for providing the tutorials!

My question is that right now, the loss is just ELBO. How do I add additional loss terms besides ELBO? For example, I might want to add an L2 loss term like `loss += lambda * F.mse_loss(x, y)`.

I guess I can probably do the normal procedure:
  `optimizer.zero_grad()`
  `loss.backward()`
  `optimizer.step()`
in addition to `svi.step()`, but I'm not sure whether this is appropriate and what `svi.step()` does exactly.

**Comment by fritzo** [comment]
Hi @jthsieh, currently the best way to specify an additional loss term is to subclass `ELBO`:
```py
class MyCustomElbo(Trace_ELBO):
    def _extra_term(self, *args, **kwargs):
        # ...your code here...
    def loss(self, *args, **kwargs):
        extra_term  = self._extra_term(*args, **kwargs)
        base_loss = super(MyCustomElbo, self).loss(*args, **kwargs)
        return base_loss + extra_term.item()
    def loss_and_grads(self, *args, **kwargs):
        extra_term  = self._extra_term(*args, **kwargs)
        extra_term.backward(create_graph=True)
        base_loss = super(MyCustomElbo, self).loss_and_grads(*args, **kwargs)
        return base_loss + extra_term.item()
```
We're looking for ways to make this easier, and we're open to any ideas you may have.

**Comment by eb8680** [comment]
I've made this easier in #1060.  We'll add an example of a custom loss in the upcoming release.

**Comment by jthsieh** [comment]
Another problem is that right now for `svi.step()`, I can't (or I don't know how to) update specific parameters. For example in GANs, I need to update parameters in the discriminator and generator separately. I think for now the easiest and most intuitive way for me is to still do the standard PyTorch `.backward()` and `optimizer.step()` for the additional losses. Several questions about this:

- Will doing so mess up the SVI algorithm?
- `svi.step()` already does a forward pass. Do I need to do it again, or is there a way I can get some variables in `model()` and `guide()` such that I can calculate the loss directly from it and do a gradient step?
- I'll need to have two optimizers, `pyro.optim.Adam` for `SVI`, and `torch.optim.Adam` for the other losses. I feel like it makes more sense to use the same optimizer. Is there a way around this?
